### PR TITLE
Enable pointer-generator T5 models in BeamSearch

### LINKED ISSUE #23123: Support pointer-generator in BeamSearch op
Support pointer-generator network in T5 implementation of BeamSearch

### PR DESCRIPTION
**Pull request by amancini-N**
### Description

Introduces a new optional input (encoder_ibnput_ids) in the decoder graph of the T5 implementation for BeamSearch. This allows usage of pointer generator networks in decoder graph.

### Motivation and Context

- Fixes #23123

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc [hunk 1]
```diff
@@ -49,11 +49,12 @@ namespace transformers {
...
                                    const std::vector<const NodeArg*>& subgraph_outputs) {
-  bool has_hidden_state = subgraph_inputs[2]->Name() == "encoder_hidden_states" ? true : false;
-  SetPastInputIndex(has_hidden_state);
+  bool has_encoder_input_ids = subgraph_inputs[1]->Name() == "encoder_input_ids";
+  bool has_hidden_state = subgraph_inputs[2 + has_encoder_input_ids]->Name() == "encoder_hidden_states";
+  SetPastInputIndex(has_hidden_state, has_encoder_input_ids);
 
-  ORT_RETURN_IF(first_past_input_index_ != 2 && first_past_input_index_ != 3,
-                "kFirstPastInputIndex currently only supports 2 or 3");
+  ORT_RETURN_IF(first_past_input_index_ != 2 && first_past_input_index_ != 3 && first_past_input_index_ != 4,
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:56]: From SetPastInputIndex implementation, this assertion of first_past_input_index_ seems always True so we can remove it.

Comment (amancini-N) [reply onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:56]: sure, will do

**Comment by amancini-N** [comment]
@tianleiwu I don't think I got which is the problem on the iOS failure. All the involved tests seems passing there. Do you have some insights?

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc [hunk 2]
```diff
@@ -75,13 +76,22 @@ Status T5DecoderSubgraph::Validate(const std::vector<const NodeArg*>& subgraph_i
...
-  ORT_RETURN_IF(subgraph_inputs[1]->Name() != "encoder_attention_mask",
-                "decoder subgraph input 1 shall be named as encoder_attention_mask, got: ",
-                subgraph_inputs[1]->Name());
-  if (first_past_input_index_ == 3) {
-    ORT_RETURN_IF(subgraph_inputs[2]->Name() != "encoder_hidden_states",
-                  "decoder subgraph input 2 shall be named as encoder_hidden_states, got: ",
-                  subgraph_inputs[2]->Name());
+  const int enc_attn_mask_index = 1 + has_encoder_input_ids_;
+  const int enc_hidden_state_index = enc_attn_mask_index + 1;
+  if (has_encoder_input_ids_) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:81]: remove this check since `has_encoder_input_ids = subgraph_inputs[1]->Name() == "encoder_input_ids"` so this will never trigger failure.

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc [hunk 3]
```diff
@@ -49,11 +49,12 @@ namespace transformers {
 
 Status T5DecoderSubgraph::Validate(const std::vector<const NodeArg*>& subgraph_inputs,
                                    const std::vector<const NodeArg*>& subgraph_outputs) {
-  bool has_hidden_state = subgraph_inputs[2]->Name() == "encoder_hidden_states" ? true : false;
-  SetPastInputIndex(has_hidden_state);
+  bool has_encoder_input_ids = subgraph_inputs[1]->Name() == "encoder_input_ids";
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:52]: Recommend to add a comment about example inputs:
input_ids, encoder_input_ids (optional), encoder_attention_mask, encoder_hidden_states (optional),
past_self_key_0, past_self_value_0, past_cross_key_0, past_cross_value_0,
past_self_key_1, past_self_value_1, past_cross_key_1, past_cross_value_1

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc [hunk 4]
```diff
@@ -238,7 +268,7 @@ Status T5DecoderSubgraph::CreateInitialFeeds(
   // When first_past_input_index_ == 3, the encoder_hidden_states and past states are copied from the second output
   // of encoder.
   // When first_past_input_index_ == 2, the past states are copied from the second output of encoder.
-  for (size_t j = static_cast<size_t>(4) - first_past_input_index_; j < encoder_fetches.size(); j++) {
+  for (size_t j = static_cast<size_t>(2) - has_hidden_state_; j < encoder_fetches.size(); j++) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:271]: The inputs `input_ids, encoder_input_ids (optional), encoder_attention_mask, encoder_hidden_states (optional), past_self_key_0, past_self_value_0, past_cross_key_0, past_cross_value_0`. 

This loop is used to add feeds for `encoder_hidden_states (optional), past_self_key_0, past_self_value_0, past_cross_key_0, past_cross_value_0`.

The encoder output is like `logits, encoder_hidden_states (optional), past_self_value_0, past_cross_key_0, past_cross_value_0, ..." so j shall starts from 2. 

Here we assume that, if encoder hidden state is not used in decoder, we shall not output it in encoder for best performance.

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc (hunk 2 above)
Comment (amancini-N) [reply onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:81]: ah right, definitely

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc (hunk 3 above)
Comment (amancini-N) [reply onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:52]: updating the comment above

**Review by tianleiwu — APPROVED** [review]
